metric_metadata/cached/cached.go

```diff
@@ -15,68 +15,117 @@
+	Expiry  time.Time    // The time at which the cache entry expires
+	Stale   time.Time    // The time at which the cache entry becomes stale
+
+	sync.Mutex // Synchronizing mutex
+
+	inflight bool           // Indicates a request is already in flight
+	enqueued bool           // Indicates a request has been enqueued
+	wg       sync.WaitGroup // Synchronizing wait group
+
+	fetchError error // Fetch error from the last attempt
 }
 
 // NewMetricMetadataAPI creates a cached API given configuration and an underlying API object.
 func NewMetricMetadataAPI(apiInstance metadata.MetricAPI, config Config) *CachedMetricMetadataAPI {
 	requests := make(chan func(metadata.Context) error, config.RequestLimit)
+	if config.Freshness == 0 {
+		config.Freshness = config.TimeToLive
+	}
 	return &CachedMetricMetadataAPI{
 		metricMetadataAPI: apiInstance,
-		getAllTagsCache:   map[api.MetricKey]CachedTagSetList{},
+		clock:             util.RealClock{},
+		getAllTagsCache:   map[api.MetricKey]*CachedTagSetList{},
+		freshness:         config.Freshness,
 		timeToLive:        config.TimeToLive,
 		backgroundQueue:   requests,
 	}
 }
 
-// CachedTagSetList is an item in the cache.
-type CachedTagSetList struct {
-	TagSets []api.TagSet // The tagsets for this metric
-	Expiry  time.Time    // The time at which the cache entry expires
-}
+// addBackgroundGetAllTagsRequest adds a job to update the lag list for the given
+// metric. Requires the client hold the lock for the item in the cache
+func (c *CachedMetricMetadataAPI) addBackgroundGetAllTagsRequest(metricKey api.MetricKey) {
+	c.queueMutex.Lock()
+	defer c.queueMutex.Unlock()
 
-// Expired tells whether the entry has zero-time (meaning absent) or is out-of-date.
-func (c CachedTagSetList) Expired() bool {
-	return c.Expiry.IsZero() || c.Expiry.Before(time.Now())
-}
+	if cap(c.backgroundQueue) <= len(c.backgroundQueue) {
+		log.Warningf("Unable to enqueue a background GetAllTags lookup for %s due to a full queue", metricKey)
+		return
+	}
 
-func (c *CachedMetricMetadataAPI) addBackgroundGetAllTagsRequest(metric api.MetricKey) {
-	c.mutex.Lock()
-	defer c.mutex.Unlock()
-	if len(c.backgroundQueue) < cap(c.backgroundQueue) {
-		// Add this function to the background queue.
-		c.backgroundQueue <- func(context metadata.Context) error {
-			defer context.Profiler.Record("CachedMetricMetadataAPI_BackgroundAction_GetAllTags")()
-			startTime := time.Now()
-			tagsets, err := c.metricMetadataAPI.GetAllTags(metric, context)
-			if err != nil {
-				return err
-			}
-			c.replaceCachedTagSet(metric, tagsets, startTime)
-			return nil
-		}
+	// No need to lock on getAllTagsCache since we already should hold the lock for
+	// the given item, so it should have already been written in a thread-safe way
+	item, ok := c.getAllTagsCache[metricKey]
+	if !ok {
+		log.Errorf("Asked to perform a background GetAllTags lookup for %s but missing entry", metricKey)
+		return
+	}
+
+	if item.enqueued {
```

Comment:
We should acquire the mutex on `item` here, instead of below, BEFORE we read `enqueued` and `inflight`, instead of after.

However, this makes unlocking it take a bit more work. I'd suggest that `fetchAndUpdateCachedTagSet` not return with the item's lock in an unlocked state (it's still allowed to unlock/lock again, just don't unlock when returning). Then we can just trivially `defer` the unlock here, (and in the other place that it's called).

To clarify: `fetchAndUpdateCachedTagSet` can still do its unlocking/locking, it just leaves things in a locked state instead of an unlocked one.